enable hypothesis testing to output a random number of tensors

hyuen opened the pull request:
Summary: we needed a random number of tensors with the same shape to write some tests, enhancing hypothesis to allow for this

Differential Revision: D16823566

hyuen commented:
@pytorchbot please retest

hyuen commented:
@pytorchbot retest this please

yinghai commented on caffe2/python/hypothesis_test_util.py:
```diff
@@ -270,6 +270,15 @@ def tensors(n, min_dim=1, max_dim=4, dtype=np.float32, elements=None, **kwargs):
 
 
+def ntensors(min_n, max_n, min_dim=1, max_dim=4, dtype=np.float32, elements=None, **kwargs):
+    dims_ = st.lists(dims(**kwargs), min_size=min_dim, max_size=max_dim)
```
Please add docstring on what this function is about.

hyuen replied:
updated

hyuen commented:
@pytorchbot please retest

yinghai commented:
Can you check CI failures?

salexspb commented:
could you add example usage? I see only new code and it is not used anywhere. Hard to see if it is working or not. If you are doing a stack, could you use gh stack or manually show usage example in the diff description?

hyuen commented:
@yinghai 
windows failure: failed to clean the workspace (environmental)

the other three tests: _dr.ci, caffe2_onnx, pytorch_linux, they all fail because:
data_parallel_model_test.py::RecurrentNetworkParallelTest::test_equiv_recurrent failed

I don't think this change is dependent on it

@salexspb for use cases this is to test a sum of random numbers of tensors, the existing infrastructure only allows us to do fixed numbers of tensors

hyuen commented on caffe2/python/hypothesis_test_util.py:
```diff
@@ -270,6 +270,19 @@ def tensors(n, min_dim=1, max_dim=4, dtype=np.float32, elements=None, **kwargs):
+def ntensors(min_n, max_n, min_dim=1, max_dim=4, dtype=np.float32, elements=None, **kwargs):
+    """
+    Create between min_n to max_n tensors, all of them with the same dimensions
+    which will be between [min_dim, max_dim] of type dtype.
```
added

Commit:
enable hypothesis testing to output a random number of tensors (#24416)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/24416

we needed a random number of tensors with the same shape to write some tests, enhancing hypothesis to allow for this

Test Plan: sum was tested this way

Differential Revision: D16823566

fbshipit-source-id: 0a0cc1606a403b7c1fc9a654a270da4adb7c8df0

hyuen commented on caffe2/python/hypothesis_test_util.py:
```diff
@@ -270,6 +270,19 @@ def tensors(n, min_dim=1, max_dim=4, dtype=np.float32, elements=None, **kwargs):
+    Create between min_n to max_n tensors, all of them with the same dimensions
+    which will be between [min_dim, max_dim] of type dtype.
+    """
+    dims_ = st.lists(dims(**kwargs), min_size=min_dim, max_size=max_dim)
```
they are unbounded, but typically we test with dimensions 1 to 4, that's how all caffe2 code is tested
